service/controller/resource/terminateunhealthynode/create.go

```diff
@@ -0,0 +1,178 @@
+	cr, err := key.ToCluster(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// check for annotation enabling the node auto repair feature
+	if _, ok := cr.Annotations[annotation.NodeTerminateUnhealthy]; !ok {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "node auto repair is not enabled for this cluster, cancelling")
+		return nil
+	}
+
+	var tenantClusterK8sClient client.Client
+	{
+		tenantClusterK8sClient, err = r.getTenantClusterClient(ctx, &cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var detectorService *detector.Detector
+	{
+		detectorConfig := detector.Config{
+			K8sClient: tenantClusterK8sClient,
+			Logger:    r.logger,
+
+			NotReadyTickThreshold: nodeTerminationTickThreshold,
+		}
+
+		detectorService, err = detector.NewDetector(detectorConfig)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	nodesToTerminate, err := detectorService.DetectBadNodes(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if len(nodesToTerminate) > 0 {
+		for _, n := range nodesToTerminate {
+			err := r.terminateNode(ctx, n, cr, tenantClusterK8sClient)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+
+		// reset tick counters on all nodes in cluster to have a graceful period after terminating nodes
+		err := detectorService.ResetTickCounters(ctx)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "level", "debug", "message", "resetting tick node counters on all nodes in tenant cluster")
+	}
+
+	return nil
+}
+
+func (r *Resource) getTenantClusterClient(ctx context.Context, cluster *capiv1alpha3.Cluster) (client.Client, error) {
+	var k8sClient k8sclient.Interface
+	{
+		restConfig, err := r.tenantRestConfigProvider.NewRestConfig(ctx, key.ClusterID(cluster), cluster.Spec.ControlPlaneEndpoint.Host)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		k8sClient, err = k8sclient.NewClients(k8sclient.ClientsConfig{
+			Logger:     r.logger,
+			RestConfig: rest.CopyConfig(restConfig),
+			SchemeBuilder: k8sclient.SchemeBuilder{
+				capzexpv1alpha3.AddToScheme,
+			},
+		})
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	return k8sClient.CtrlClient(), nil
+}
+
+func (r *Resource) terminateNode(ctx context.Context, node corev1.Node, cluster capiv1alpha3.Cluster, ctrlClient client.Client) error {
+	if node.Labels["role"] != "worker" {
+		return microerror.Maskf(unsupportedOperationError, "Termination of master nodes is not supported on Azure")
+	}
+
+	vmssClient, err := r.azureClientsFactory.GetVirtualMachineScaleSetsClient(ctx, cluster.ObjectMeta)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// Scale VMSS up by one.
+	var vmssName string
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "Retrieving AzureMachinePool CR")
+		amp := capzexpv1alpha3.AzureMachinePool{}
+		err := ctrlClient.Get(ctx, client.ObjectKey{Namespace: cluster.Namespace, Name: node.Labels[label.MachinePool]}, &amp)
```

Comment:
Here you get the error for unknown kind because the passed `ctrlClient` is the client to tenant cluster which doesn't have CRDs for our types.

Reply:
thanks